gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsUtils.java

```diff
@@ -0,0 +1,178 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.rpc.mtls;
+
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.SecurityUtils;
+import com.google.api.core.BetaApi;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.List;
+
+/** Utilities for mutual TLS. */
+@BetaApi
+public class MtlsUtils {
+  @VisibleForTesting
+  static class DefaultMtlsProvider implements MtlsProvider {
+    private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+        System.getProperty("user.home") + "/.secureConnect/context_aware_metadata.json";
+
+    /** GOOGLE_API_USE_CLIENT_CERTIFICATE environment variable. */
+    public static final String GOOGLE_API_USE_CLIENT_CERTIFICATE =
+        "GOOGLE_API_USE_CLIENT_CERTIFICATE";
+
+    /** GOOGLE_API_USE_MTLS_ENDPOINT environment variable. */
+    public static final String GOOGLE_API_USE_MTLS_ENDPOINT = "GOOGLE_API_USE_MTLS_ENDPOINT";
+
+    interface EnvironmentProvider {
+      String getenv(String name);
+    }
+
+    static class SystemEnvironmentProvider implements EnvironmentProvider {
+      @Override
+      public String getenv(String name) {
+        return System.getenv(name);
+      }
+    }
+
+    DefaultMtlsProvider() {
+      this(new SystemEnvironmentProvider(), DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+    }
+
+    private EnvironmentProvider envProvider;
+    private String metadataPath;
+
+    /** Cache the key store so we don't create a new one every time getKeyStore is called. */
+    private KeyStore keyStore = null;
+
+    @VisibleForTesting
+    DefaultMtlsProvider(EnvironmentProvider envProvider, String metadataPath) {
+      this.envProvider = envProvider;
+      this.metadataPath = metadataPath;
+    }
+
+    @Override
+    public boolean useMtlsClientCertificate() {
+      String useClientCertificate = envProvider.getenv(GOOGLE_API_USE_CLIENT_CERTIFICATE);
+      return "true".equals(useClientCertificate);
+    }
+
+    @Override
+    public String useMtlsEndpoint() {
+      return envProvider.getenv(GOOGLE_API_USE_MTLS_ENDPOINT);
+    }
+
+    @Override
+    public String getKeyStorePassword() {
+      return "";
+    }
+
+    @Override
+    public KeyStore getKeyStore() throws IOException, GeneralSecurityException {
+      if (keyStore != null) {
```

Comment:
It seems like lazy initialization approach is used here and it also tries to prevent reloading the keystore data if it has already been loaded. Please strongly consider making this class thread safe in this case to make sure that keyStore is not attempted to be initialized more than once if simultaneously executed from multiple threads, plus to make sure that the changes are visible from the other threads (i.e. at least make `keyStore` volatile)

Reply:
Now the code doesn't do lazy initialization so we don't have to worry about it any more. Creating a key store is cheap, also we just create it once when client is created, so we don't need to re use it.